[13.0] split module sale_order_line_packaging_qty

grindtildeath opened the pull request:
PR #1174 did introduce multiple features into the same module (`sale_order_line_packaging_qty`) which had a well defined name and scope.
This PR aims to split the different features that were merged in `sale_order_line_packaging_qty` into a different module `sale_by_packaging` to provide well packaged features among modules.

- sale_order_line_packaging_qty

This module adds a "Package quantity" field on sale order line in order to
define a quantity according to the selected "Package".


- sale_by_packaging

This module provides different configuration option to manage packagings on
sale orders.

Following options are available to define which packaging type can be sold and
which product can only be sold by packaging.

* Can be sold: On product packaging type model, this checkbox defines if product
  packagings from this particular type are available to be selected on sale
  order line.

* Sell only by packaging: On product template model, this checkbox restricts
  sales of these products if no packaging is selected on the sale order line.
  If no packaging is selected, it will either be auto-assigned if the quantity
  on the sale order line matches a packaging quantity or an error will be raised.

grindtildeath committed:
Revert "[IMP][sale_order_line_packaging_qty] Sell product / unit or packaging"

This reverts commit 5eece31b79d5c6fec1c316fc7f6fb5679222f1c4.

grindtildeath committed:
Create module product_packaging_can_be_sold

grindtildeath committed:
Create module product_packaging_type_can_be_sold

grindtildeath committed:
Create module sale_product_by_packaging_only

grindtildeath committed:
Create module sale_product_packaging_restriction

guewen commented on sale_product_by_packaging_only/models/sale_order_line.py:
```diff
@@ -0,0 +1,92 @@
+            ):
+                pack_multiple = line.product_id._which_pack_multiple(product_uom_qty)
+                if pack_multiple:
+                    vals.update({"product_packaging": pack_multiple.id})
```
Would be worth to extract the behavior added in `write` in another method: easier to customize and extend this behavior if it's in another method compared to being in the middle of a write chain.

guewen commented on sale_product_by_packaging_only/models/sale_order_line.py:
```diff
@@ -0,0 +1,92 @@
+        if product and product.sell_only_by_packaging and not product_packaging:
+            pack_multiple = product._which_pack_multiple(product_uom_qty)
+            if pack_multiple:
+                vals.update({"product_packaging": pack_multiple.id})
```
same remark as for the write above

grindtildeath commented:
ping @Tonow-c2c @jgrandguillaume @simahawk since you contributed or reviewed #1174

guewen commented:
As discussed, we should keep sale_order_line_packaging_qty for the quantity only, reflecting what's done in `purchase_order_line_packaging_qty`, and have a single module `sale_product_by_packaging` that depends on the qty one for the other use cases (packaging & packaging type can be sold + sell product by packaging only + filter list of available packaging).

grindtildeath committed:
Create module sale_by_packaging

The functions of this module were first added into sale_order_line_packaging_qty.

After sorting out and extracting the different functions in multiple modules in
the previous commits, this one gets it all together in a single module.

jgrandguillaume reviewed (approved):
That looks good and in-line with the purchase part. Thx for the job !

simahawk commented on sale_by_packaging/models/sale_order_line.py:
```diff
@@ -0,0 +1,160 @@
+            return super().write(vals)
+        for line in self:
+            line_vals = vals.copy()
+            line_vals.update(self._write_auto_assign_packaging(line_vals))
```
shouldn't be called on line?
```suggestion
            line_vals.update(line._write_auto_assign_packaging(line_vals))
```

simahawk commented on sale_by_packaging/models/sale_order_line.py:
```diff
@@ -0,0 +1,160 @@
+                self.product_id, self.product_uom_qty, self.product_uom
+            ):
+                warning_msg = {
+                    "title": _("Product quantity can not be packed"),
```
```suggestion
                    "title": _("Product quantity cannot be packed"),
```

simahawk commented on sale_by_packaging/tests/test_sale_only_by_packaging.py:
```diff
@@ -0,0 +1,114 @@
+        )
+
+    def test_onchange_qty_is_pack_multiple(self):
+        order = self.env["sale.order"].create({"partner_id": self.partner.id})
```
the order can be created in setupClass only I think

simahawk commented on sale_by_packaging/tests/test_sale_only_by_packaging.py:
```diff
@@ -0,0 +1,114 @@
+            }
+        )
+        self.assertTrue(order_line_1.product_packaging)
+        # self.assertTrue(order_line_1.product_packaging_qty)
```
any dead code to cleanup? :)

simahawk reviewed (changes requested):
@grindtildeath good work, tnx! Can you squash when done w/ my comments?

simahawk commented on sale_by_packaging/models/sale_order_line.py:
```diff
@@ -0,0 +1,160 @@
+            self.update(
+                {
+                    "product_packaging": first_packaging.id,
+                    "product_uom_qty": first_packaging.qty,
```
here you could have ZERO as qty. Can we do anything better? Eg: not setting the qty and / or show a warning?

grindtildeath replied:
I filtered out packagings with a qty of zero and added a note in roadmap :wink:

grindtildeath committed:
Consider packagings with a qty of zero

simahawk commented:
/ocabot merge minor

i-vyshnevska commented on sale_by_packaging/models/product_template.py:
```diff
@@ -0,0 +1,62 @@
+        "packaging defined",
+    )
+
+    @api.constrains("sell_only_by_packaging", "sale_ok", "packaging_ids")
```
do you need a "packaging_ids" field here?

grindtildeath replied:
Not needed indeed, good catch
